### ISSUE TITLE
Sync with crowdsec api in stream mode is inconsistent 

### ISSUE THREAD

**Issue by cRUSHr2012** [crowdsecurity/cs-wordpress-bouncer#97]
Hi, 

 I am using Crowdsec Bouncer Plugin 1.4.3 with WP 5.93 and PHP 7.4.29.
I am using a single Redis DB with multiple WP sites. One of them is set to sync at 60 seconds (WP cron is replaced by a system cron job, and wp-config disables CRON), the rest of them have > 60000 seconds interval.
I am manually banning IP's and monitoring **/var/log/crowdsec_api.log**, **/var/log/crowdsec.log**, **redis-cli monitor** and the the **access log** for the website that syncs at each 60sec.

_**Problem : sometimes the IP is not added to Redis, so they are not banned.**_

How do I test:
I ban the IP 81.196.86.95 with **cscli decisions add --ip 81.196.86.95 --duration 5m**.
I see the API log entry :

_time="24-05-2022 12:05:36" level=info msg="(441fca0e645b4b41b4c2d6da56663f0c3xIHLlSvv2snSDYT/cscli) manual 'ban' from '441fca0e645b4b41b4c2d6da56663f0c3xIHLlSvv2snSDYT' by ip 81.196.86.95 : 5m ban on Ip 81.196.86.95"_

I can also see it in : **cscli decisions list** table
I have **redis-cli** monitor command running. If I click **Refresh the cache now**, I see the Redis entry being added:
Redis 10 DB is used in all configs.

_1653383978.667692 [10 127.0.0.1:57260] "MGET" "cacheConfig"
1653383978.730722 [10 127.0.0.1:57260] "MGET" "SXA6ODEuMTk2Ljg2Ljk1"
1653383978.731144 [10 127.0.0.1:57260] "SETEX" "SXA6ODEuMTk2Ljg2Ljk1" "296" "\x00\x00\x00\x02\x14\x01\x06\x00\x14\x04\x06\x00\x11\x03ban\x06\x01\nb\x8c\xa4S\x06\x02\n\x00'\xb6\xd1\x06\x03\x06\x00"_

If I try to browse from the above IP, i get a 403 as expected.

If I add the next IP 81.196.86.96, and wait for the cron to run, in the access.log I see entries at each 60s:

_84.247.81.123 - - [24/May/2022:12:07:01 +0300] "GET /wp-cron.php?doing_wp_cron HTTP/2.0" 200 20 "-" "Wget/1.21"
84.247.81.123 - - [24/May/2022:12:08:01 +0300] "GET /wp-cron.php?doing_wp_cron HTTP/2.0" 200 20 "-" "Wget/1.21"
84.247.81.123 - - [24/May/2022:12:09:01 +0300] "GET /wp-cron.php?doing_wp_cron HTTP/2.0" 200 20 "-" "Wget/1.21"
84.247.81.123 - - [24/May/2022:12:10:02 +0300] "GET /wp-cron.php?doing_wp_cron HTTP/2.0" 200 20 "-" "Wget/1.21"_

In **crowdsec_api.log** I see the requests from the plugin:

_time="24-05-2022 12:07:01" level=info msg="127.0.0.1 - [Tue, 24 May 2022 12:07:01 EEST] \"GET /v1/decisions/stream?startup=false&scopes=Ip%2CRange HTTP/1.0 200 67.459313ms \"WordPress CrowdSec Bouncer/v1.4.3\" \""
time="24-05-2022 12:08:01" level=info msg="127.0.0.1 - [Tue, 24 May 2022 12:08:01 EEST] \"GET /v1/decisions/stream?startup=false&scopes=Ip%2CRange HTTP/1.0 200 151.400157ms \"WordPress CrowdSec Bouncer/v1.4.3\" \""
time="24-05-2022 12:09:02" level=info msg="127.0.0.1 - [Tue, 24 May 2022 12:09:02 EEST] \"GET /v1/decisions/stream?startup=false&scopes=Ip%2CRange HTTP/1.0 200 151.97765ms \"WordPress CrowdSec Bouncer/v1.4.3\" \""
time="24-05-2022 12:10:02" level=info msg="127.0.0.1 - [Tue, 24 May 2022 12:10:02 EEST] \"GET /v1/decisions/stream?startup=false&scopes=Ip%2CRange HTTP/1.0 200 190.641162ms \"WordPress CrowdSec Bouncer/v1.4.3\" \""_

Redis-cli monitor doesn't show any SETEX entries.

The timestamps match, this means **wp-cron** triggers the refresh from the API.

If i wait and add other IP's each minute, 81.196.86.97 / 81.196.86.98 / 81.196.86.99 / 81.196.86.100 it doesn't (sometimes) pick them up. The logs show the API being queried but that's it.
Manually clicking the **Refresh the cache now** works and I recorded the SETEX value for each, to manually do a MGET after them.

81.196.86.95 	-	SXA6ODEuMTk2Ljg2Ljk1
81.196.86.96	-	SXA6ODEuMTk2Ljg2Ljk2
81.196.86.97	-	SXA6ODEuMTk2Ljg2Ljk3
81.196.86.98	-	SXA6ODEuMTk2Ljg2Ljk4
81.196.86.99	-	SXA6ODEuMTk2Ljg2Ljk5

I noticed that If I add **81.196.86.97** and is not picked up, after the cron runs, then I add **81.196.86.98**, and manually click **Refresh the cache now**, only **81.196.86.98 / SXA6ODEuMTk2Ljg2Ljk4** key is added to Redis. Checking for the key **SXA6ODEuMTk2Ljg2Ljk3** corresponding to **81.196.86.97** shows (nil).

In the decisions table both IP's are still present, but only the last one is added in Redis. I've tried with longer ban periods (so that the TTL for the key does not expire too soon for the tests), but it's the same. I've made sure that the IP's are not in the Crowsec DB already (blacklisted from central API). I have no other websites quering the API (both LIVE / STREAM mode).

I think the latest decisions are always synced from the API, but the IP's aren't always added to Redis.

This behaviour seems to happen when unbanning IP's also.
I need a way to debug the API refresh done by the plugin.

**Comment by julienloizelet** [comment]
Hi,
thanks for this detailed issue.

> I need a way to debug the API refresh done by the plugin.

Well, there is also a debug.log file (if you enable the option in the plugin) that you will find in the plugins sources :  `wp-content/plugins/cs-wordpress-bouncer/logs` but I guess it won't help us very much : you should find some `CACHE_UPDATED` log message, but we have seen that we cannot trust the `deleted` count, and maybe the `new` count...  (I created an issue about this bug).

So, for now, the only thing I see to debug this is to look in the source code directly.

**We will work on it as soon as possible**, but here is what I can say "briefly":

First : it's quite strange that 
> Manually clicking the Refresh the cache now works

As far as I understand the source code, the cron job or a manual click are doing exactly the same thing : 

- for the cron job, we can look at the `crowdSecRefreshBlocklist` function in the `inc/scheduling.php` (line 19)
- for a manual click, we can look at `refreshBouncerCacheInAdminPage` in the `inc/admin/init.php` (line 61)
These 2 functions are doing the following: 
- `$settings = getDatabaseSettings();` : we retrieve the settings of the crowdsec plugin
- `$bouncer = getBouncerInstance($settings);` : we instantiate a bouncer with those settings
- `$bouncer->refreshBlocklistCache();` : we call a method that is intended to "refresh the cache`

Secondly, we have to look at this `refreshBlocklistCache()` (see `vendor/crowdsec/bouncer/src/Bouncer.php` (line 186).
This method just calls the `pullUpdates` function (see `vendor/crowdsec/bouncer/src/ApiCache.php` (line 614).
The `pullUpdates` function starts by retrieving the "fresh" decisions by calling the `/v1/decisions/stream?startup=false` url that you can see in your `crowdsec_api.log` (`$this->apiClient->getStreamedDecisions(false, $this->getScopes());`). And then, it should remove the found deleted decisions (`$this->removeRemediations($deletedDecisions)` ) and save the new ones ( `$this->saveRemediations($newDecisions)`)

I guess there are at least 2 possibilities that should cause your issue:
1) for some reason, the crowdsec api itself does not return what we expect : i.e the `getStreamedDecisions` result is wrong
2) for some reason, the `pullUpdates` method does not save or delete correctly the  `getStreamedDecisions` result.

> I think the latest decisions are always synced from the API, but the IP's aren't always added to Redis.

If your intuition is correct, we'll have to investigate on the option 2); especially on the `saveRemediations` method (see `vendor/crowdsec/bouncer/src/ApiCache.php` line 394)  that should add each new decision to the cache.

**Comment by julienloizelet** [comment]
> Is the API query done by the plugin, the same in both LIVE / STREAM mode

No, the `getStreamedDecisions` call is only used in stream mode.

And indeed, https://github.com/crowdsecurity/cs-wordpress-bouncer/issues/96#issuecomment-1126959013 and https://github.com/crowdsecurity/cs-wordpress-bouncer/issues/96#issuecomment-1127307219 are maybe related to your issue.

**Comment by cRUSHr2012** [comment]
I've connected a PHP8 site running a cPANEL to the crowdsec API / Redis.
In redis-cli monitor the log is a bit different.

From PHP 8.0:

`1653466503.117885 [10 84.247.81.166:48428] "SETEX" "SXA6ODEuMTk2Ljg2Ljkw" "244" "a:1:{i:0;a:4:{i:0;s:3:\"ban\";i:1;i:1653466748;i:2;i:2768057;i:3;i:0;}}"`

From PHP 7.4:

`1653162950.679117 [10 127.0.0.1:40530] "SETEX" "SXA6ODEuMTk2Ljg2Ljkw" "14394" "\x00\x00\x00\x02\x14\x01\x06\x00\x14\x04\x06\x00\x11\x03ban\x06\x01\nb\x89|\x01\x06\x02\n\x00!e\xf1\x06\x03\x06\x00"`

I will test from PHP8 / cPANEL Wordpress.

**Comment by cRUSHr2012** [comment]
It seems there is no difference if I use PHP8 

I manually banned 81.196.86.93 at 12:14:14.
From the logs automatically an IP is also added at 12:14:49.

Crowdsec API log:

```
==> /var/log/crowdsec.log <==
time="25-05-2022 12:14:14" level=info msg="(441fca0e645b4b41b4c2d6da56663f0c3xIHLlSvv2snSDYT/cscli) manual 'ban' from '441fca0e645b4b41b4c2d6da56663f0c3xIHLlSvv2snSDYT' by ip 81.196.86.93 : 5m ban on Ip 81.196.86.93"
time="25-05-2022 12:14:49" level=info msg="Ip 154.54.249.203 performed 'crowdsecurity/http-bad-user-agent' (2 events over 15.288614805s) at 2022-05-25 09:14:49.069303867 +0000 UTC"
time="25-05-2022 12:14:50" level=info msg="(441fca0e645b4b41b4c2d6da56663f0c3xIHLlSvv2snSDYT/crowdsec) crowdsecurity/http-bad-user-agent by ip 154.54.249.203 (US/210743) : 4h ban on Ip 154.54.249.203"
```

The API was queried by the plugin at 12:15:03

`time="25-05-2022 12:15:03" level=info msg="84.247.81.166 - [Wed, 25 May 2022 12:15:03 EEST] \"GET /v1/decisions/stream?startup=false&scopes=Ip%2CRange HTTP/1.1 200 202.108182ms \"WordPress CrowdSec Bouncer/v1.4.3\" \""`

Redis-cli monitor only shows:

```
1653470103.132086 [0 84.247.81.166:54298] "SELECT" "10"
1653470103.146146 [10 84.247.81.166:54298] "MGET" "cacheConfig"
```

1653470103.132086 is 12:15:03 the same when the API was queried by the plugin.

The plugin debug.log (timestamp is +3h) :

```
2022-05-25T09:15:03.146176+00:00|100|{"type":"API_CACHE_INIT","adapter":"Symfony\\Component\\Cache\\Adapter\\RedisAdapter","mode":"stream","exp_clean_ips":1,"exp_bad_ips":1,"warmed_up":"true","geolocation":{"save_in_session":true,"enabled":false,"type":"maxmind","maxmind":{"database_type":"country"}}}
2022-05-25T09:15:03.146415+00:00|100|{"type":"REST_CLIENT_INIT","base_uri":"http://84.247.81.123:8888","timeout":1}
2022-05-25T09:15:03.146475+00:00|100|{"type":"API_CLIENT_INIT","user_agent":"WordPress CrowdSec Bouncer/v1.4.3"}
2022-05-25T09:15:03.146523+00:00|100|{"type":"START_CACHE_UPDATE"}
2022-05-25T09:15:03.146591+00:00|100|{"type":"HTTP CALL","method":"GET","uri":"http://84.247.81.123:8888/v1/decisions/stream?startup=false&scopes=Ip%2CRange","content":null}
2022-05-25T09:15:03.351466+00:00|100|{"type":"CACHE_UPDATED","deleted":0,"new":0}
```

**Comment by cRUSHr2012** [comment]
Update : the IP **154.54.249.203** was picked up later, after repeated entries in the API (?)
I don't know about : 111.162.142.12. 
The IP's are naturally banned from the logs. 

**crowdsec.log**

```
time="25-05-2022 12:14:14" level=info msg="(441fca0e645b4b41b4c2d6da56663f0c3xIHLlSvv2snSDYT/cscli) manual 'ban' from '441fca0e645b4b41b4c2d6da56663f0c3xIHLlSvv2snSDYT' by ip 81.196.86.93 : 5m ban on Ip 81.196.86.93"
time="25-05-2022 12:14:50" level=info msg="(441fca0e645b4b41b4c2d6da56663f0c3xIHLlSvv2snSDYT/crowdsec) crowdsecurity/http-bad-user-agent by ip 154.54.249.203 (US/210743) : 4h ban on Ip 154.54.249.203"
time="25-05-2022 12:16:23" level=info msg="(441fca0e645b4b41b4c2d6da56663f0c3xIHLlSvv2snSDYT/crowdsec) crowdsecurity/http-open-proxy by ip 111.162.142.12 (CN/4837) : 4h ban on Ip 111.162.142.12"
time="25-05-2022 12:17:09" level=info msg="(441fca0e645b4b41b4c2d6da56663f0c3xIHLlSvv2snSDYT/crowdsec) crowdsecurity/http-bad-user-agent by ip 154.54.249.203 (US/210743) : 4h ban on Ip 154.54.249.203"
time="25-05-2022 12:19:22" level=info msg="(441fca0e645b4b41b4c2d6da56663f0c3xIHLlSvv2snSDYT/crowdsec) crowdsecurity/http-bad-user-agent by ip 154.54.249.203 (US/210743) : 4h ban on Ip 154.54.249.203"
time="25-05-2022 12:21:42" level=info msg="(441fca0e645b4b41b4c2d6da56663f0c3xIHLlSvv2snSDYT/crowdsec) crowdsecurity/http-bad-user-agent by ip 154.54.249.203 (US/210743) : 4h ban on Ip 154.54.249.203"
time="25-05-2022 12:23:56" level=info msg="(441fca0e645b4b41b4c2d6da56663f0c3xIHLlSvv2snSDYT/crowdsec) crowdsecurity/http-bad-user-agent by ip 154.54.249.203 (US/210743) : 4h ban on Ip 154.54.249.203"
time="25-05-2022 12:25:59" level=info msg="(441fca0e645b4b41b4c2d6da56663f0c3xIHLlSvv2snSDYT/crowdsec) crowdsecurity/http-bad-user-agent by ip 154.54.249.203 (US/210743) : 4h ban on Ip 154.54.249.203"
time="25-05-2022 12:28:25" level=info msg="(441fca0e645b4b41b4c2d6da56663f0c3xIHLlSvv2snSDYT/crowdsec) crowdsecurity/http-bad-user-agent by ip 154.54.249.203 (US/210743) : 4h ban on Ip 154.54.249.203"
time="25-05-2022 12:30:27" level=info msg="(441fca0e645b4b41b4c2d6da56663f0c3xIHLlSvv2snSDYT/crowdsec) crowdsecurity/http-bad-user-agent by ip 154.54.249.203 (US/210743) : 4h ban on Ip 154.54.249.203"
time="25-05-2022 12:32:50" level=info msg="(441fca0e645b4b41b4c2d6da56663f0c3xIHLlSvv2snSDYT/crowdsec) crowdsecurity/http-bad-user-agent by ip 154.54.249.203 (US/210743) : 4h ban on Ip 154.54.249.203"
time="25-05-2022 12:35:07" level=info msg="(441fca0e645b4b41b4c2d6da56663f0c3xIHLlSvv2snSDYT/crowdsec) crowdsecurity/http-bad-user-agent by ip 154.54.249.203 (US/210743) : 4h ban on Ip 154.54.249.203"
time="25-05-2022 12:37:36" level=info msg="(441fca0e645b4b41b4c2d6da56663f0c3xIHLlSvv2snSDYT/crowdsec) crowdsecurity/http-bad-user-agent by ip 154.54.249.203 (US/210743) : 4h ban on Ip 154.54.249.203"
```

plugin **debug.log**

`cat debug-2022-05-25.log | grep "\"new\":1}"`

```
2022-05-25T09:11:02.564354+00:00|100|{"type":"CACHE_UPDATED","deleted":0,"new":1}
2022-05-25T09:13:02.939351+00:00|100|{"type":"CACHE_UPDATED","deleted":0,"new":1}
2022-05-25T09:17:03.209429+00:00|100|{"type":"CACHE_UPDATED","deleted":0,"new":1}
2022-05-25T09:18:03.060734+00:00|100|{"type":"CACHE_UPDATED","deleted":0,"new":1}
2022-05-25T09:20:02.863562+00:00|100|{"type":"CACHE_UPDATED","deleted":0,"new":1}
2022-05-25T09:24:02.257737+00:00|100|{"type":"CACHE_UPDATED","deleted":0,"new":1}
2022-05-25T09:29:02.174023+00:00|100|{"type":"CACHE_UPDATED","deleted":0,"new":1}
2022-05-25T09:36:02.307047+00:00|100|{"type":"CACHE_UPDATED","deleted":0,"new":1}
```

The date is the same on both servers.
Do you know how the hash key is generated for the IP value ?  If I somehow miss the redis-cli log command I want to manually check the IP in REDIS.

I'll do more tests today.

**Comment by julienloizelet** [comment]
Hi @cRUSHr2012 ,
I've got 2 news: 

1) There is a bug with the CrowdSec agent itself:
> We have indeed spotted a bug in versions 1.3.2 up to and including 1.3.4 that causes a bug with the stream mode but only if the refresh rate is greater than 60s (so if you refresh with a delay longer than 1 minute)

Here is the opened issue: https://github.com/crowdsecurity/crowdsec/issues/1565

I do not remember which version of the crowdsec agent you are using and which frequency you are using, but your issue may be related to this. Please let me know.

2)  I told you in another issue:
> if you refresh on the first WP, a refresh on the second will not retrieve the decisions that have been updated for the first one

(see https://github.com/crowdsecurity/cs-wordpress-bouncer/issues/96#issuecomment-1127780748)
Following this, we decided to enable refresh cron job only in one Wordpress (the other wordpress have a big number to "disable" the cron).

I apologize, but It seems that I was wrong. There is a way to pull fresh decisions for every WP : one should use a different bouncer key for each WordPress instance.

Indeed, a CrowdSec core team member told me that: 
> the Crowdsec LAPI stores the timestamp of the last pull associated with the api key and uses this to find out which IPs have been added/deleted since the last time.

Thus, I guess you could continue to use the "big number workaround" for the cron (it should not harm), but, as a good practice, please consider setting a different bouncer API key for each WordPress instance.

**Comment by cRUSHr2012** [comment]
Hi @julienloizelet 

1. The plugin version i use on all WP sites is 1.3.4. The refresh interval is 60 seconds, controlled by a system cron job. The other sites have a refresh > 60000 sec
2. If all the websites use the same Redis DB, it shouldn't be a problem if a refresh at 60000 seconds comes right before/after the one the runs at each 60 seconds. The request goes to the API, and the decisions to the same shared Redis DB. Right ?
3. To lower the refresh interval under 60 seconds I will use add second cron job, but with a sleep of 30 seconds. This should refresh the API at each 30 seconds.

Watching the **redis-cli monitor** command, there is something that I don't understand:

```
1653642443.536370 [0 127.0.0.1:51014] "SELECT" "10"
1653642443.546354 [10 127.0.0.1:51014] "MGET" "cacheConfig"
1653642443.549282 [10 127.0.0.1:51014] "EXISTS" "SXA6ODEuMTk2Ljg2Ljkw"
1653642443.549456 [10 127.0.0.1:51014] "MGET" "SXA6ODEuMTk2Ljg2Ljkw"
1653642443.549871 [10 127.0.0.1:51014] "UNLINK" "SXA6ODEuMTk2Ljg2Ljkw"
1653642449.193222 [0 127.0.0.1:51018] "SELECT" "10"
1653642449.203096 [10 127.0.0.1:51018] "MGET" "cacheConfig"
1653642449.205778 [10 127.0.0.1:51018] "EXISTS" "SXA6ODEuMTk2Ljg2Ljkw"
1653642449.205958 [10 127.0.0.1:51018] "MGET" "SXA6ODEuMTk2Ljg2Ljkw"
1653642449.206405 [10 127.0.0.1:51018] "UNLINK" "SXA6ODEuMTk2Ljg2Ljkw"
```

Why after each "EXISTS" **key** and "MGET" **key**, there is a "UNLINK" **key** command ?
I am doing normal requests from a second machine, but why is this IP checked and "deleted" from Redis, if it's not stored there (not banned) at all ? This happens for each request.

**Comment by julienloizelet** [comment]
Hi,
1.
> The plugin version i use on all WP sites is 1.3.4

I guess you are talking about the CrowdSec agent version. Well, that means that you could be affected by the "> 60 seconds" bug.
In the description of the issue, there is:
> it cause to the bouncer to miss a full 60s of decisions

and it looks like your issue.

2. 
> it shouldn't be a problem if a refresh at 60000 seconds comes right before/after the one the runs at each 60 seconds.

I think so too. By the way, setting a different key for each instance seems to be the recommended way of sharing a LAPI for multiple bouncers.

3.
This "> 60 seconds" bug  is really annoying as 1 minute interval is the minimum that we can set for a cron job...  
Your workaround should indeed do the job.


For the last part (the Redis question), I'm sorry but I  have no clue for now.

**Comment by cRUSHr2012** [comment]
Hi, 

  Indeed running 2 system cron jobs at each minute (one of them with a sleep 30 seconds command) does the trick. Also in the WP Crowdsec plugin we need to make sure to have : "Resync decisions each 30 seconds" set.

**Comment by julienloizelet** [comment]
> > The plugin version i use on all WP sites is 1.3.4
> 
> I guess you are talking about the CrowdSec agent version. Well, that means that you could be affected by the "> 60 seconds" bug. In the description of the issue, there is:
> 
> > it cause to the bouncer to miss a full 60s of decisions
> 
> and it looks like your issue.

Hi,

the  "> 60 seconds" bug has been fixed in the [Release 1.4.1](https://github.com/crowdsecurity/crowdsec/releases/tag/v1.4.1) of the CrowdSec Agent.

Thus, I guess I can close this issue.

Thanks
